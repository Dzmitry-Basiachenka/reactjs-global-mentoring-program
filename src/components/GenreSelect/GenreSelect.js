import React from 'react';

class GenreSelect extends React.Component {

  constructor(props) {
    super(props);
    this.state = {
      selectedGenres: this.props.selectedGenres
    }
  }

  render() {
    return <fieldset>
      <legend>Select Genre</legend>
      {
        this.props.genres.map((genre, key) => {
          return <div key={genre}>
            <input
              type='checkbox'
              defaultChecked={this.state.selectedGenres.includes(genre) ? 'checked' : ''}
              id={genre}
              name={genre}
              value={genre}
              onClick={this.handleClick}>
            </input>
            <label
              htmlFor={genre}>
              {genre}
            </label>
          </div>
        })
      }
    </fieldset>
  }

  handleClick = (event) => {
    const selectedGenres = [...this.state.selectedGenres];
    const selectedGenre = event.target.value;

    if (selectedGenres.includes(selectedGenre)) {
      selectedGenres.splice(selectedGenres.indexOf(selectedGenre), 1);

      this.setState(prevState => ({ ...prevState, selectedGenres: selectedGenres }));
      this.props.onSelect(selectedGenre, false);
    } else {
      selectedGenres.push(selectedGenre);

      this.setState(prevState => ({ ...prevState, selectedGenres: selectedGenres }));
      this.props.onSelect(selectedGenre, true)
    }
  }

}

export default GenreSelect;
