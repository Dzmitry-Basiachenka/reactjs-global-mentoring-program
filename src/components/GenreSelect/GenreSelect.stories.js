import GenreSelect from './GenreSelect';

export default {
  component: GenreSelect
};

export const AllUnchecked = {
  args: {
    genres: ['Documentary', 'Comedy', 'Horror', 'Crime'],
    selectedGenres: [],
    onSelect: (genre, selection) => console.log(`genre: ${genre}, selection: ${selection}`),
  },
  parameters: {
  }
};

export const AllChecked = {
  args: {
    genres: ['Documentary', 'Comedy', 'Horror', 'Crime'],
    selectedGenres: ['Documentary', 'Comedy', 'Horror', 'Crime'],
    onSelect: (genre, selection) => console.log(`genre: ${genre}, selection: ${selection}`),
  },
  parameters: {
  }
};
