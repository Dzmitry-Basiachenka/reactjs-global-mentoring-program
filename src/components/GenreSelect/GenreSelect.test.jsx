import React from 'react';
import renderer from 'react-test-renderer';
import { cleanup, render, screen, fireEvent } from '@testing-library/react';

import GenreSelect from './GenreSelect';

describe('GenreSelect', () => {

  afterEach(cleanup);

  it('should match snapshot', () => {
    const genreSelect = renderer
      .create(
        <GenreSelect
          genres={['Documentary', 'Comedy', 'Horror', 'Crime']}
          selectedGenres={['Documentary', 'Comedy']}
        />
      )
      .toJSON();

    expect(genreSelect).toMatchSnapshot();
  });

  // Test that component renders all genres passed in props
  it('should render correctly', () => {
    const genres = ['Documentary', 'Comedy', 'Horror', 'Crime'];
    const selectedGenres = ['Documentary', 'Comedy'];

    render(
      <GenreSelect
        genres={genres}
        selectedGenres={selectedGenres}
      />
    );

    const genreLabels = screen.getAllByRole('label')
      .map(label => label.innerHTML);
    expect(genreLabels).toEqual(genres);
  });

  // Test that component highlights a selected genre passed in props
  it('should highlight a selected genre', () => {
    const genres = ['Documentary', 'Comedy', 'Horror', 'Crime'];
    const selectedGenres = ['Documentary', 'Comedy'];

    render(
      <GenreSelect
        genres={genres}
        selectedGenres={selectedGenres}
        onSelect={console.log}
      />
    );

    const genreCheckboxes = screen.getAllByRole('checkbox')
      .filter(checkbox => checkbox.checked)
      .map(checkbox => checkbox.value);
    expect(genreCheckboxes).toEqual(selectedGenres);
  });

  // Test that after a click event on a genre button component calls "onChange" callback and passes correct genre in arguments
  it('should call the handler', () => {
    const genres = ['Documentary', 'Comedy', 'Horror', 'Crime'];
    const selectedGenres = ['Documentary', 'Comedy'];
    const handleSelectMock = jest.fn();

    render(
      <GenreSelect
        genres={genres}
        selectedGenres={selectedGenres}
        onSelect={handleSelectMock}
      />
    );

    const deselectingGenre = genres[1];
    const deselectingCheckbox = screen.getByText(deselectingGenre);
    fireEvent.click(deselectingCheckbox, { target: { innerText: deselectingCheckbox.innerHTML } });
    expect(handleSelectMock).toHaveBeenCalledWith(deselectingGenre, false);

    const selectingGenre = genres[2];
    const selectingCheckbox = screen.getByText(selectingGenre);
    fireEvent.click(selectingCheckbox, { target: { innerText: selectingCheckbox.innerHTML } });
    expect(handleSelectMock).toHaveBeenCalledWith(selectingGenre, true);
  });

});
