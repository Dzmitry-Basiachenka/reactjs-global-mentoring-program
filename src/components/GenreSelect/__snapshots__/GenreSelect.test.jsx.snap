// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GenreSelect should match snapshot 1`] = `
<fieldset>
  <legend>
    Select Genre
  </legend>
  <div>
    <input
      defaultChecked="checked"
      id="Documentary"
      name="Documentary"
      onClick={[Function]}
      type="checkbox"
      value="Documentary"
    />
    <label
      htmlFor="Documentary"
      role="label"
    >
      Documentary
    </label>
  </div>
  <div>
    <input
      defaultChecked="checked"
      id="Comedy"
      name="Comedy"
      onClick={[Function]}
      type="checkbox"
      value="Comedy"
    />
    <label
      htmlFor="Comedy"
      role="label"
    >
      Comedy
    </label>
  </div>
  <div>
    <input
      defaultChecked=""
      id="Horror"
      name="Horror"
      onClick={[Function]}
      type="checkbox"
      value="Horror"
    />
    <label
      htmlFor="Horror"
      role="label"
    >
      Horror
    </label>
  </div>
  <div>
    <input
      defaultChecked=""
      id="Crime"
      name="Crime"
      onClick={[Function]}
      type="checkbox"
      value="Crime"
    />
    <label
      htmlFor="Crime"
      role="label"
    >
      Crime
    </label>
  </div>
</fieldset>
`;
